src/examples/libpmemobj/libart/libart2/art.h

```diff
@@ -0,0 +1,213 @@
+ *
+ *         Author:  Andreas Bluemle, Dieter Kasper
+ *                  andreas.bluemle@itxperts.de
+ *                  dieter.kasper@ts.fujitsu.com
+ *
+ *   Organization:  FUJITSU TECHNOLOGY SOLUTIONS GMBH
+ *
+ * ===========================================================================
+ */
+
+/*
+ * based on https://github.com/armon/libart/src/art.h
+ */
+
+#ifndef _ART_H
+#define _ART_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define ART_VARIANT "libart on libpmemobj"
+#define ART_VERSION "2.0"
+
+#define MAX_PREFIX_LEN 10
+
+typedef enum {
+	NODE4		= 0,
+	NODE16		= 1,
+	NODE48		= 2,
+	NODE256		= 3,
+	art_leaf_t	= 4,
+	art_node_types	= 5   /* number of different art_nodes */
+} art_node_type;
+
+extern char *art_node_names[];
+
+/*
+ * forward declarations; these are required when typedef shall be
+ * used instead of struct
+ */
+struct _art_node;	typedef struct _art_node art_node;
+struct _art_node4;	typedef struct _art_node4 art_node4;
+struct _art_node16;	typedef struct _art_node16 art_node16;
+struct _art_node48;	typedef struct _art_node48 art_node48;
+struct _art_node256;	typedef struct _art_node256 art_node256;
+struct _art_leaf;	typedef struct _art_leaf art_leaf;
+
+extern size_t art_tree_root_type_num;
+extern size_t art_node4_type_num;
+extern size_t art_node16_type_num;
+extern size_t art_node48_type_num;
+extern size_t art_node256_type_num;
+extern size_t art_leaf_type_num;
+
+POBJ_LAYOUT_BEGIN(arttree_tx);
+POBJ_LAYOUT_ROOT(arttree_tx, struct art_tree_root);
+POBJ_LAYOUT_TOID(arttree_tx, art_node4);
+POBJ_LAYOUT_TOID(arttree_tx, art_node16);
+POBJ_LAYOUT_TOID(arttree_tx, art_node48);
+POBJ_LAYOUT_TOID(arttree_tx, art_node256);
+POBJ_LAYOUT_TOID(arttree_tx, art_leaf);
+POBJ_LAYOUT_END(arttree_tx);
+
+/*
+ * This struct is included as part of all the various node sizes
+ */
+struct _art_node {
+	uint8_t num_children;
+	uint32_t partial_len;
+	unsigned char partial[MAX_PREFIX_LEN];
+};
+
+/*
+ * Small node with only 4 children
+ */
+struct _art_node4 {
+	art_node n;
+	unsigned char keys[4];
+	PMEMoid children[4];
+};
+
+/*
+ * Node with 16 children
+ */
+struct _art_node16 {
+	art_node n;
+	unsigned char keys[16];
+	PMEMoid children[16];
+};
+
+/*
+ * Node with 48 children, but a full 256 byte field.
+ */
+struct _art_node48 {
+	art_node n;
+	unsigned char keys[256];
+	PMEMoid children[48];
+};
+
+/*
+ * Full node with 256 children
+ */
+struct _art_node256 {
+	art_node n;
+	PMEMoid children[256];
+};
+
+/*
+ * Represents a leaf. These are of arbitrary size, as they include the key.
+ */
+struct _art_leaf {
+	size_t key_len;
+	size_t val_len;
+	size_t buffer_len;
+	unsigned char buffer[0];	/* single buffer for key + value */
+};
+
+struct art_tree_root {
+	size_t size;
+	PMEMoid root;
+};
+
+typedef struct _cb_data {
+	PMEMoid node;
+	int child_idx;
+} cb_data;
+
+/*
+ * Macros to manipulate art_node tags
+ */
+#define IS_LEAF(oid) (OID_INSTANCE_OF(oid) == art_leaf_type_num))
```

Comment:
libart2 does no longer use the type safety macros from libpmemobj. It does type checking explicitely.